qiskit/pulse/filters.py

```diff
@@ -13,15 +13,18 @@
 """A collection of functions that filter instructions in a pulse program."""
 
 import abc
+from functools import singledispatch
 from typing import Callable, List, Union, Iterable, Optional, Tuple, Any
 
 import numpy as np
 
-from qiskit.pulse import Schedule
+from qiskit.pulse import Schedule, ScheduleBlock, Instruction
 from qiskit.pulse.channels import Channel
 from qiskit.pulse.schedule import Interval
+from qiskit.pulse.exceptions import PulseError
 
 
+@singledispatch
 def filter_instructions(
```

Comment:
Can you implement the dispatcher in the following form for better readability and robustness? Same for other functions.
```python
@singledispatch
def some_function(arg, ...):  # when arg doesn't match any registered types, fall into this.
    raise TypeError(f"type(arg) is not valid data format as an input to some_function.")

@some_function.register
def do_type1(arg: Type1, ....):  # instead of underscore, we can give human readable name.
    ...

@some_function.register
def do_type2(arg: Type2, ....):
    ...
```

Reply:
This is a smart idea! I updated following this rule 👉 https://github.com/Qiskit/qiskit-terra/pull/9971/commits/a2757ce9e43ccbf34468a9d8d78fdad181285063

Note that I adopted `def handle_type1(...):` instead of `def do_type1(...):`, since `handle` sounds more direct, but I can change it to `def do_type1(...):` if you wish.